### PR TITLE
[iOS] Changes in Border clipping part

### LINKED ISSUE #14927: iOS: Content in `Border` with `StrokeShape` of `Ellipse` not rendered correctly.
### Description

I attempted to add font icon content into a circular "button-like" component I was building;

The component is a border with an ellipse stroke shape, with a label containing the font icon in it.

On Android this renders correctly;
android

On iOS it is quite broken;
ios

It is as if the ellipse stroke clip is being applied to the children of `Border` as well as the border itself.

### Steps to Reproduce

1. Open my repro project.
2. Run in iOS
3. Observe issue.

### Link to public reproduction project repository

https://github.com/RobTF/mauiplay/tree/master/BorderStrokeShapeIssue

### Version with bug

7.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

iOS

### Affected platform versions

iOS 16

### Did you find any workaround?

Avoid the issue by making nothing a child of `Border` if at all possible, e.g.

```xaml
            <Border
                Style="{StaticResource ButtonStyle}">
                <Label
                    Style="{StaticResource ButtonIcon}"
                    Text="circle"
                    TextColor="#ff3333"
                    FontFamily="FontAwesomeSolid" />
            </Border>
```
becomes...
```xaml
            <Grid>
                <Border Style="{StaticResource ButtonStyle}" />
                <Label
                    Style="{StaticResource ButtonIcon}"
                    Text="circle"
                    TextColor="#ff3333"
                    FontFamily="FontAwesomeSolid" />                
            </Grid>
```

### Relevant log output

```shell
None.
```

### LINKED ISSUE #15183: [regression/8.0] [iOS] Border appears to be offsetting content incorrectly
### Description

```XAML

 <Grid>
                <Border StrokeThickness="5" Stroke="Red">
                    <Border.StrokeShape>
                        <RoundRectangle CornerRadius="5" />
                    </Border.StrokeShape>
                    <Grid BackgroundColor="Blue" HeightRequest="40">
                        <Label Text="Control in a Grid" TextColor="White"/>
                    </Grid>
                </Border>
            </Grid>

            <ContentView Background="Green" BackgroundColor="Green">
                <Border StrokeThickness="5" Stroke="Red">
                    <Border.StrokeShape>
                        <RoundRectangle CornerRadius="5" />
                    </Border.StrokeShape>
                    <Grid BackgroundColor="Blue" HeightRequest="40">
                        <Label Text="Control in a ContentView" TextColor="White"/>
                    </Grid>
                </Border>
            </ContentView>

            <!-- Control in a grid is rendered properly -->
            <Grid>
                <Border StrokeThickness="5" Stroke="Red">
                    <Border.StrokeShape>
                        <RoundRectangle CornerRadius="5" />
                    </Border.StrokeShape>
                    <Grid BackgroundColor="Blue" HeightRequest="40">
                        <Label Text="Control in a Grid" TextColor="White"/>
                    </Grid>
                </Border>
            </Grid>
```

### NET7.0
image

### NET8.0
image

### ANDROID NET8.0
image

### LINKED ISSUE #15794: Border background color on Android is different than iOS
### Description

net8 pre5 - I'm seeing Android change in the background color of Border.

Screenshot 2023-06-21 at 3 16 14 PM

### Steps to Reproduce

```
using CommunityToolkit.Maui.Markup;
using Microsoft.Maui.Controls.Shapes;
using static CommunityToolkit.Maui.Markup.GridRowsColumns;

namespace ControlGallery.Pages;

public class BorderPage : BaseContentPage
{
    protected override void Build()
    {
        this.Title = "Borders";

        this.Content = new ScrollView{
            Content = new VerticalStackLayout{
                Children = {
                    new H1("Rectangle"),
                    new Separator(),
                    new Border{
                        StrokeShape = new Rectangle(),
                        Content = new Label().Text("Basic Rectangle").Margin(12,8)
                    },

                    new H1("Round Rectangle"),
                    new Separator(),
                    new Border{
                        StrokeShape = new RoundRectangle{ CornerRadius = 20},
                        Content = new Label().Text("Round Rectangle").Margin(12,8)
                    },

                    new H1("Border Thickness"),
                    new Separator(),
                    new Border{
                        StrokeThickness = 8,
                        StrokeShape = new Rectangle(),
                        Content = new Label().Text("StrokeThickness=8").Margin(12,8)
                    },

                    new H1("StrokeDashOffset"),
                    new Separator(),
                    new Border{
                        StrokeDashOffset = 9,
                        StrokeDashArray = new DoubleCollection{1,1},
                        StrokeShape = new Rectangle(),
                        Content = new Label().Text("StrokeDashOffset=8, StrokeDashArray=1,1").Margin(12,8)
                    },

                    new H1("StrokeLineJoin"),
                    new Separator(),
                    new Border{
                        StrokeLineJoin = PenLineJoin.Bevel,
                        StrokeDashOffset = 9,
                        StrokeDashArray = new DoubleCollection{1,1},
                        StrokeShape = new Rectangle(),
                        Content = new Label().Text("StrokeLineJoine=Bevel").Margin(12,8)
                    },

                    new H1("StrokeLineCap"),
                    new Separator(),
                    new Border{
                        StrokeLineCap = PenLineCap.Flat,
                        StrokeDashOffset = 9,
                        StrokeDashArray = new DoubleCollection{1,1},
                        StrokeShape = new Rectangle(),
                        Content = new Label().Text("StrokeLineCap=Flat").Margin(12,8)
                    },

                    new H1("Ellipse"),
                    new Separator(),
                    new Grid
                    {
                        ColumnDefinitions = Columns.Define(78, Star, Star),
                        RowDefinitions = Rows.Define(48,30),
                        ColumnSpacing = 20,
                        RowSpacing = 0,
                        Children =
                        {
                            new Border{
                                StrokeShape = new Ellipse(),
                                Stroke = Colors.Red,
                                StrokeThickness = 6,
                                Content = new Label().Text("Hey").Center().FontSize(24)
                            }
                            .Row(0).Column(0).RowSpan(2)
                            .Background(Colors.Azure)
                            .Center()
                            .Size(72)
                            
                        }
                    },
                }
            }.DynamicResource(Label.StyleProperty, "MainContainer")
        };
    }
}
```

Global style

```
<Style TargetType="Border">
        <Setter Property="Stroke" Value="{AppThemeBinding Light={StaticResource Gray200}, Dark={StaticResource Gray900}}" />
        <Setter Property="StrokeShape" Value="Rectangle"/>
        <Setter Property="StrokeThickness" Value="1"/>
    </Style>
```

### Link to public reproduction project repository

https://github.com/davidortinau/ControlGallery/blob/sharpination/src/ControlGallery/Pages/Controls/BorderPage.cs

### Version with bug

8.0.0-preview.5.8529

### Last version that worked well

7.0.86

### Affected platforms

Android

### Affected platform versions

33

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by jsuarezruiz**
### Description of Change

Changes in Border clipping part:
- Added validation to check if it is necessary to clip the content or not. If the content does not overflow, unnecessary clipping is avoided.
- The clip is now made from the proportional part that overflows. 

fix-ios-issue
_NOTE: In the previous gif, the red color is the BackgroundColor from the Label inside the Border._

To test the changes can use the added sample under the Border gallery from the .NET MAUI Gallery App.

### Issues Fixed

Fixes #15794
Fixes #14927
Fixes #15183

### CODE REVIEW

File: src/Core/src/Platform/iOS/ContentView.cs [hunk 1]
```diff
@@ -114,9 +136,28 @@ void SetClip()
 			else
 				path = clipShape?.PathForBounds(bounds);
 
-			var nativePath = path?.AsCGPath();
+			var platformClipPath = path?.AsCGPath();
+
+			return platformClipPath;
+		}
+
+		bool RequireClip(CGPath? platformClipPath)
```
Comment (jsuarezruiz) [inline src/Core/src/Platform/iOS/ContentView.cs:144]: We only apply the clip if the content overflows the border.

**Review by jstedfast — DISMISSED** [review]

**Comment by jsuarezruiz** [comment]
Added new sample to validate 15183:
Captura de pantalla 2023-07-05 a las 17 47 55

File: src/Core/src/Handlers/Border/BorderHandler.iOS.cs [hunk 2]
```diff
@@ -1,12 +1,19 @@
...
 
 namespace Microsoft.Maui.Handlers
 {
 	public partial class BorderHandler : ViewHandler<IBorderView, ContentView>
 	{
+		WeakReference<LayoutView>? _container;
+
+		~BorderHandler()
+		{
+			Container?.Dispose();
```
Comment (PureWeen) [inline src/Core/src/Handlers/Border/BorderHandler.iOS.cs:13]: Currently these are crashing the Device Tests

`at Microsoft.Maui.Handlers.BorderHandler.Finalize() in /Users/runner/work/1/s/src/Core/src/Handlers/Border/BorderHandler.iOS.cs:line 14`

I don't think we want to call `Dispose` on these. I'm thinking there's something else inside this PR that's causing things to leak.

Comment (PureWeen) [inline src/Core/src/Handlers/Border/BorderHandler.iOS.cs:13]: Currently these are crashing the Device Tests

`at Microsoft.Maui.Handlers.BorderHandler.Finalize() in /Users/runner/work/1/s/src/Core/src/Handlers/Border/BorderHandler.iOS.cs:line 14`

I don't think we want to call `Dispose` on these. I'm thinking there's something else inside this PR that's causing things to leak.

**Review by PureWeen — CHANGES_REQUESTED** [review]

**Review by jonathanpeppers — COMMENTED** [review]
Just to check it's not the test:

https://github.com/dotnet/maui/blob/bcffd04ca71f9cbca7c21075fd58b2745d92fdf4/src/Controls/tests/DeviceTests/Elements/Border/BorderTests.cs#L111-L119

More recently, I've been moving to this pattern:

https://github.com/dotnet/maui/blob/bcffd04ca71f9cbca7c21075fd58b2745d92fdf4/src/Controls/tests/DeviceTests/Elements/NavigationPage/NavigationPageTests.cs#L331-L336

But if it still fails, it probably is a circular reference.

I wrote this doc on Friday: https://github.com/dotnet/maui/wiki/Memory-Leaks

Can you follow the steps on there to record a `.gcdump` file on Catalyst after the test fails?

**Comment by jonathanpeppers** [comment]
One of the merges from `main` picked up this change: https://github.com/dotnet/maui/commit/9ab334665ee4b2e0dd18da9fd4be41e6476d50c6

So, the `Border` memory leak test passes now. There isn't a leak here.

Unfortunately, the entire Controls device tests hang indefinitely. Running locally, there looks like there an infinite loop involving this line:

image

**Review by jonathanpeppers — DISMISSED** [review]
LGTM now if all the tests pass. 👍

**Review by PureWeen — CHANGES_REQUESTED** [review]
@jstedfast is going to review more thoroughly. 

A couple questions we have from our review call

- Does it make sense to be passing the `BorderHandler.VirtualView` as the `CrossplatformLayout` to both the inner `LayoutView` and the outer `ContentView`

https://github.com/dotnet/maui/blob/cf7a81431aa561cdba56153cceb954e24dae4556/src/Core/src/Handlers/Border/BorderHandler.iOS.cs#L18
https://github.com/dotnet/maui/blob/cf7a81431aa561cdba56153cceb954e24dae4556/src/Core/src/Handlers/Border/BorderHandler.iOS.cs#L78

It makes it so `UIView.LayoutSubviews` from the outer and inner view both get routed to `Border.cs` but `GetDesiredSize` only triggers on the inner view. That just feels a bit odd to me. 


- Instead of having the `ContainerView` on the handler itself and thus having to cast the `IBorderHandler` can that instead be its own controls `BorderContainerView`
  - The other question I have here is why we are storing the `ContainerView` at all on the handler? Can we just set that as the subview?

**Review by jstedfast — DISMISSED** [review]

File: src/Core/src/Platform/iOS/ContentView.cs (hunk 1 above)
Comment (rmarinho) [reply src/Core/src/Platform/iOS/ContentView.cs:146]: I think we need to add this as a comment here.

File: src/Core/src/Platform/iOS/ContentView.cs [hunk 3]
```diff
@@ -21,10 +21,19 @@ public override void LayoutSubviews()
...
-			var bounds = AdjustForSafeArea(Bounds).ToRectangle();
+			if (Clip is not null)
+			{
+				var bounds = AdjustForSafeArea(Bounds).ToRectangle();
+
+				if (Subviews.Length != 0)
+				{
+					if (Subviews[0] is LayoutView content && content.Frame != bounds)
+						content.Frame = bounds;
+				}
```
Comment (rmarinho) [inline src/Core/src/Platform/iOS/ContentView.cs:32]:
```suggestion
				if (Subviews.FirstOrDefaultNoLinq() is LayoutView content && content.Frame != bounds)
				{
						content.Frame = bounds;
				}
```

File: src/Core/src/Platform/iOS/ContentView.cs [hunk 4]
```diff
@@ -103,8 +112,24 @@ void SetClip()
 			// we use the double to get the correct value. Here, again, we use the double to get the correct clip shape size values.
 			var strokeWidth = 2 * strokeThickness;
 
-			var bounds = new RectF(0, 0, (float)frame.Width - strokeWidth, (float)frame.Height - strokeWidth);
+			var bounds = new RectF(strokeWidth / 2, strokeWidth / 2, (float)frame.Width - strokeWidth, (float)frame.Height - strokeWidth);
```
Comment (rmarinho) [inline src/Core/src/Platform/iOS/ContentView.cs:115]: Can you add a comment here explain what's happening to the bounds ?

For example why isn't the Width `(float)frame.Width - (strokeWidth /2)` ? And why are we clipping just half of the stroke?!

File: src/Controls/tests/DeviceTests/Elements/Border/BorderTests.cs [hunk 5]
```diff
@@ -110,11 +110,12 @@ public async Task DoesNotLeak()
 			});
 
 			await AssertionExtensions.WaitForGC(handlerReference, platformViewReference);
+
 			Assert.False(handlerReference.IsAlive, "Handler should not be alive!");
 			Assert.False(platformViewReference.IsAlive, "PlatformView should not be alive!");
 		}
 
-		/* Commented out for now until border issues are fixed
+#if IOS
```
Comment (dotMorten) [inline src/Controls/tests/DeviceTests/Elements/Border/BorderTests.cs:118]: Should there be a comment linking to the issues that prevents it from running on the other platforms?

**Review by hartez — DISMISSED** [review]
Putting a hold on this, it needs some work before it can be merged.

**Review by PureWeen — CHANGES_REQUESTED** [review]
The clip from the border doesn't appear to be updating

- border gallery
- Border Playground
- increase top left corner radius

image


if you change the content from Image then back to label it clips.

image


But then it stays clipped

image


Border Clipped Playground is another good page for repro


image


Then if you update the BorderWidth

image

**Comment by hartez** [comment]
> The clip from the border doesn't appear to be updating
> 
> * border gallery
> * Border Playground
> * increase top left corner radius
> 

Try it now.

**Review by PureWeen — APPROVED** [review]

**Comment by ederbond** [comment]
This must be backported to .NET 7 ASAP.

**Pull request merged**